core-compose/src/main/java/com/squareup/workflow/ui/compose/RenderAsState.kt

```diff
@@ -0,0 +1,244 @@
+ * limitations under the License.
+ */
+@file:Suppress("NOTHING_TO_INLINE")
+
+package com.squareup.workflow.ui.compose
+
+import androidx.annotation.VisibleForTesting
+import androidx.compose.Composable
+import androidx.compose.CompositionLifecycleObserver
+import androidx.compose.FrameManager
+import androidx.compose.MutableState
+import androidx.compose.State
+import androidx.compose.mutableStateOf
+import androidx.compose.remember
+import androidx.ui.core.CoroutineContextAmbient
+import androidx.ui.core.Ref
+import androidx.ui.savedinstancestate.Saver
+import androidx.ui.savedinstancestate.SaverScope
+import androidx.ui.savedinstancestate.UiSavedStateRegistryAmbient
+import androidx.ui.savedinstancestate.savedInstanceState
+import com.squareup.workflow.Snapshot
+import com.squareup.workflow.Workflow
+import com.squareup.workflow.diagnostic.WorkflowDiagnosticListener
+import com.squareup.workflow.launchWorkflowIn
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import okio.ByteString
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Runs this [Workflow] as long as this composable is part of the composition, and returns a
+ * [State] object that will be updated whenever the runtime emits a new [RenderingT].
+ *
+ * The workflow runtime will be started when this function is first added to the composition, and
+ * cancelled when it is removed. The first rendering will be available immediately as soon as this
+ * function returns, as [State.value]. Composables that read this value will automatically recompose
+ * whenever the runtime emits a new rendering.
+ *
+ * [Snapshot]s from the runtime will automatically be saved to the current
+ * [UiSavedStateRegistry][androidx.ui.savedinstancestate.UiSavedStateRegistry]. When the runtime is
+ * started, if a snapshot exists in the registry, it will be used to restore the workflows.
+ *
+ * @receiver The [Workflow] to run. If the value of the receiver changes to a different [Workflow]
+ * while this function is in the composition, the runtime will be restarted with the new workflow.
+ * @param props The [PropsT] for the root [Workflow]. Changes to this value across different
+ * compositions will cause the root workflow to re-render with the new props.
+ * @param onOutput A function that will be executed whenever the root [Workflow] emits an output.
+ * @param diagnosticListener An optional [WorkflowDiagnosticListener] to start the runtime with. If
+ * this value changes while this function is in the composition, the runtime will be restarted.
+ */
+@Composable
+fun <PropsT, OutputT : Any, RenderingT> Workflow<PropsT, OutputT, RenderingT>.renderAsState(
```

Comment:
This is so clean. It has that "this is obviously the right thing to do, why did we do anything else" feel to it.

Reply:
Yep. It's almost the exact same API as Flow/Rx subscriptions in Compose too.